javascript/organizing_your_javascript_code/webpack.md

```diff
@@ -1,84 +1,420 @@
+// greeting.js
+export const greeting = "Hello, Odinite!";
+```
+
+Great, now we have an `index.js` that imports from, and so depends on, `greeting.js`. In order to bundle this, we'll also want a Webpack configuration file, which will contain all the details we need for bundling, such as the entry point, the output destination, and anything like plugins and loaders (which we will cover shortly).
+
+Back in your project root (so outside of `src`), create a `webpack.config.js` file that contains the following:
+
+```javascript
+// webpack.config.js
+const path = require("path");
+
+module.exports = {
+  mode: "development",
+  entry: "./src/index.js",
+  output: {
+    filename: "main.js",
+    path: path.resolve(__dirname, "dist"),
+    clean: true,
+  },
+};
+```
+
+Yes, you may have noticed this file uses CommonJS (CJS) syntax instead of ESM. That's because this file (and Webpack itself) runs in NodeJS and not the browser. By default, NodeJS uses CJS syntax, and the configuration file also contains some CJS-specific things. We need not worry about this - this is just stuff we need for Webpack to do its thing.
+
+You'll notice the exported object contains a few key sections:
+
+- `mode`: For now, we will just leave this in development mode, as it will be more useful to us. We will revisit this and production mode in a later lesson.
+- `entry`: A file path from the config file to whichever file is our entry point, which in this case is `src/index.js`.
+- `output`: An object containing information about the output bundle.
+  - `filename`: The name of the output bundle - it can be anything you want.
+  - `path`: The path to the output directory, in this case, `dist`. If this directory doesn't already exist when we run Webpack, it will automatically create it for us as well. Don't worry too much about why we have the `path.resolve` part - this is just the way Webpack recommends we specify the output directory.
+  - `clean`: If we include this option and set it to `true`, then each time we run Webpack to bundle, it will empty the output directory first before bundling the files into it. This helps us keep `dist` clean, so it only contains the files produced by the most recent bundling.
+  
+With these files all in place, let's run Webpack and see what happens!
+
+```bash
+npx webpack
+```
+
+You should see that Webpack has created a `dist` directory for us containing a `main.js` file! Inside this file is...a lot of stuff... Don't worry, most of this stuff is just for development tools we will use later. If you go ahead and run this file with `node dist/main.js`, you should see `Hello, Odinite!` logged in the terminal.
+
+Congratulations! You've just made your first bundle with Webpack!
+
+### Handling HTML
+
+Let's actually include some HTML. After all, we're interested in making websites! Since HTML isn't JavaScript, Webpack can't just bundle it straight away, but there's a nifty tool we can use called `HtmlWebpackPlugin` that's just perfect for us.
```

Comment:
This certainly can be reinforced. Given the perspective of someone going through this for the first time, do you think a reinforcement for HTML specifically is significantly valuable given that earlier in the lesson there's a section about `src` and `dist`? Which includes the following:

> `src` is where we keep all of our website's source code, essentially where all of our work will be done (with an exception being altering any configuration files in the root of the project). When we run Webpack to bundle our code, it will output the bundled files into the `dist` directory. The idea is that if someone were to fork or clone the project, they would not need the `dist` directory, as they'd just be able to run Webpack to build from `src` into their own `dist`. Similarly, to deploy our website, we would only need the `dist` code and nothing else. Keep that in mind! Work inside `src`, build into `dist`, then deploy from there!

My idea was that this would be enough to justify not mentioning the HTML and dist thing, since it's already been introduced that all work goes in `src`, not `dist`. So I was thinking new learners might not even consider putting HTML in `dist` and working from there. But if you think this is still valuable to mention, then I can do so.

Reply:
I thought it was essential since every developer has to work with source and distribution at some point. The more curious learners who are quick to grasp things might, at some point, wonder why they're doing this src and dist thing mindlessly and not actually directly modifying dist stuff.

Reply:
Perhaps a small reminder that we're using a template in `src` because we reserve `dist` solely for the distribution code? Don't think we need to dig into it further given that the bit I highlighted above already explains the purpose and different roles of `src`/`dist`. I'm writing this with the perspective of a new learner going through this for the first time, as opposed to people who might've gotten confused with the official Webpack tutorials previously.

Reply:
Oh, thank you for the reminder. 

For some reason the live preview did not load for me until yesterday, so I was able to take a look at the updated rewrite only today. And yes I finally saw the note today that you're talking about, so my apologies. I somehow did not notice it beforehand.

Agreed, any more explanations and it could lead to confusion.

Reply:
Awesome, just wanted to make sure we're on the same page. Most of the current confusion about dist/src comes from the official webpack tutorial starting off by doing things in dist, which I've always found very confusing given that's entirely against the point of what src/dist are for and what webpack is designed to do...

But hopefully without that, people won't even consider working in dist because the new lesson never does, and (hopefully) clearly explains the purpose of those two dirs.